Always skip body for 1xx, 204, and 304 responses

azaretsky opened the pull request:
https://tools.ietf.org/html/rfc7230#section-3.3.3:
... any response with a 1xx (Informational), 204 (No Content), or 304 (Not Modified) status code is always terminated by the first empty line after the header fields, regardless of the header fields present in the message, and thus cannot contain a message body.

Fixes: https://github.com/nodejs/http-parser/issues/251

azaretsky committed:
Always skip body for 1xx, 204, and 304 responses

Fixes: https://github.com/nodejs/http-parser/issues/251

bnoordhuis commented on http_parser.c:
```diff
@@ -887,6 +885,12 @@ size_t http_parser_execute (http_parser *parser,
+        if (parser->status_code / 100 == 1 || /* 1xx e.g. Continue */
+            parser->status_code == 204 ||     /* No Content */
+            parser->status_code == 304) {     /* Not Modified */
+            parser->flags |= F_SKIPBODY;
```
Indent error, should be two spaces.

Something for a separate commit, since you're here anyway: it might be faster to write `parser->status_code / 100 == 1` as `parser->status_code >= 100 && parser_status_code < 200`.

Long division is moderately expensive and I'm not sure if compilers are smart enough to turn the first into the latter.

tolikz replied:
Thanks, corrected.

Quite surprisingly both clang 6.0.0. and gcc 7.3.0 are that smart actually (at least for x86-64) but I changed this comparison anyway.

clang:
```
        addl    $-100, %eax
        cmpl    $100, %eax
        jb      .LBB0_60
```

gcc:
```
        subl    $100, %eax
        cmpw    $99, %ax
        ja      .L170
```

tolikz replied:
Wouldn't you mind if I mark this conversation resolved?

bnoordhuis replied:
Sorry, of course! I'll resolve it.

tolikz replied:
Thanks!

bnoordhuis commented on test.c:
```diff
@@ -1848,8 +1848,7 @@ const struct message responses[] =
   ,.response_status= "Switching Protocols"
-  ,.body= "body"
-  ,.upgrade= "proto"
+  ,.upgrade= "bodyproto"
```
See #363, it's possible someone might be relying on the current behavior.

tolikz replied:
This PR only changes the tests for 101 responses. The HTTP_200_RESPONSE_WITH_UPGRADE_HEADER* tests are unaffected (and they all pass).

RFC 7230 explicitly forbids using Content-Length or Transfer-Encoding in 1xx responses ("A server MUST NOT send" etc.) and RFC 2616 also had similarly strict requirements ("All 1xx ... responses MUST NOT include a message-body"), so I think it would probably be ok.

bnoordhuis reviewed (commented):
While the change LGTM in general I'm kind of worried it might break existing http-parser consumers due to the change in behavior.

azaretsky committed:
Change the check for 1xx status code

"status_code >= 100 && status_code < 200" might be faster than
"status_code / 100 == 1"

tolikz commented:
It's quite hard to tell if this behavior change will affect anyone since it seems that these responses are rarely encountered in the wild if at all: 1xx, 204, and 304 responses without a declared message body length are already correctly handled by the parser, while 1xx and 204 with a message body are not legitimate according to the spec, and real servers often do not include content-length or transfer-encoidng in 304 responses (at least IIS, nginx and apache httpd don't).

So there's a great chance that this change is harmless but also purely cosmetic and no-one actually needs it.

bnoordhuis commented:
cc @indutny - perhaps you can weigh in?

indutny commented:
I think it might be too hard to propagate description this change to the consumers of http-parser. In order to properly do it, we'll have to do major release.

@nodejs/http : do we consider major release a possibility, or would we prefer to stick with llhttp and do majority of changes there instead?

tolikz commented:
Just for the record: I will try to keep this branch rebased on top of the current master in case anyone wants to use it before the next major release.
